internal/providers/openstack/openstack.go

```diff
@@ -166,14 +159,86 @@ func fetchConfigFromDevice(logger *log.Logger, ctx context.Context, path string)
 	return os.ReadFile(filepath.Join(mnt, configDriveUserdataPath))
 }
 
+func FindIPv6InterfaceName() (string, error) {
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return "", err
+	}
+
+	for _, iface := range interfaces {
+		// Check if the interface is up and not loopback
+		if iface.Flags&net.FlagUp != 0 && iface.Flags&net.FlagLoopback == 0 {
+			addrs, err := iface.Addrs()
+			if err != nil {
+				continue
+			}
+
+			for _, addr := range addrs {
+				// Check if the address is an IPv6
+				if ipnet, ok := addr.(*net.IPNet); ok && ipnet.IP.To4() == nil && !ipnet.IP.IsLinkLocalUnicast() {
+					return iface.Name, nil
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("no IPv6 interface name found")
+}
+
 func fetchConfigFromMetadataService(f *resource.Fetcher) ([]byte, error) {
-	res, err := f.FetchToBuffer(metadataServiceUrl, resource.FetchOptions{})
 
-	// the metadata server exists but doesn't contain any actual metadata,
-	// assume that there is no config specified
-	if err == resource.ErrNotFound {
-		return nil, nil
+	// Find IPv6 name
+	iface, err := FindIPv6InterfaceName()
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct URL
+	var (
+		ipv4MetadataServiceUrl = url.URL{
+			Scheme: "http",
+			Host:   "169.254.169.254",
+			Path:   "openstack/latest/user_data",
+		}
+		ipv6MetadataServiceUrl = url.URL{
+			Scheme: "http",
+			Host:   fmt.Sprintf("[fe80::a9fe:a9fe%%%s]", url.PathEscape(iface)),
```

Comment:
Maybe we need to change to `[fe80::a9fe:a9fe%25%s]` refer to https://docs.openstack.org/nova/latest/user/metadata.html#the-metadata-service